pkg/loghttp/entry.go

```diff
@@ -18,6 +18,7 @@ func init() {
 type Entry struct {
 	Timestamp time.Time
 	Line      string
+	Labels    string
```

Comment:
Making this a string feels counterintuitive compared to using `map[string]string`, which is what we use via the [`LabelSet`](https://github.com/grafana/loki/blob/main/pkg/loghttp/labels.go#L21-L22) struct used in our v1 push payloads in [`loghttp/push`](https://github.com/grafana/loki/blob/main/pkg/loghttp/push/push.go#L104). Point being, I think we should use JSON directly (not json with embedded stringified json fields) when `application/json` is the `Content-Type`. When snappy-encoding is used, `logproto` can use the string variants for consistency with how it treats labels.

Reply:
The problem I see with making this a map, and keeping as a string the labels for an entry in logproto, is that converting from `loghttp.Entry` to `logproto.Entry` is going to slower.

Right now (both with string labels), we just cast the pointer for the entries array of a stream since they have the same memory layout.
https://github.com/grafana/loki/blob/44b0932247909e5d196087b88ec15ae00ad7254d/pkg/loghttp/query.go#L236

If we decide to make one a map and another one a string, we won't be able to do that cast anymore (different memory layouts). Hence, we'll need to go entry by entry converting it from `loghttp.Entry` to `logproto.Entry`.

IMO, we should make both use the same layout.

Reply:
I implemented this in https://github.com/grafana/loki/pull/9694/commits/f43dfab7febf8784e6624ea3459826ac0ec0309b

This is what we need to do instead of the cast:
https://github.com/grafana/loki/blob/f43dfab7febf8784e6624ea3459826ac0ec0309b/pkg/loghttp/query.go#L236-L239

If we decide to keep this as a string, we can revert the commit. If we decide to make logproto.Entry to use a map for labels, we can build on top of it.

Reply:
according to our API docs, we use JSON object for stream labels )https://grafana.com/docs/loki/latest/api/#push-log-entries-to-loki
so, it makes sense to use JSON object for line metadata also… but you are right @salvacorts that there is no need to convert it to the string when we convert `loghttp.Entry` to `logproto.Entry` ….

Reply:
@salvacorts You're right, making this anything but a `string` here will result in performance losses because this it would need to parse labels potentially on _every entry_ coming back from subqueries. Since this happens before we filter out unnecessary entries (over `limit`, etc), this can quickly result in higher resource utilization, OOMs, etc. 

Now, we have a few options:
1) Keep this as a `string` in memory, but turn it into a `map` when return the `json` response from the http api. This will maintain performance (no conversion from protos) & correctly return a map via the json api.
2) Use `LabelSet` for this field (stored as a `map`). This requires parsing strings -> json for potentially _every_ log line returned from queriers (yikes!). Even when combined with (4), this would require transforming `slice -> map`.
3) Use `labels.Labels` (stored as a `slice`). When combined with (4) this incurs no transformation cost and the resulting conversion to the json map for the external API is simple & relatively inexpensive.
4) Due to the performance implications here, I've changed my opinion on making [`EntryAdapter.labels`](https://github.com/grafana/loki/blob/main/pkg/push/push.proto#L33) a `string` for parity with [`StreamAdapter.labels`](https://github.com/grafana/loki/blob/main/pkg/push/push.proto#L24). Instead, I think we should use the [`LabelAdapter`](https://github.com/grafana/loki/blob/main/pkg/logproto/logproto.proto#L337) pattern to send the labels in the protos as a slice and combine it with (3) to not need to pay conversion costs.

Does this make sense?

Reply:
Agree. Implemented it in https://github.com/grafana/loki/pull/9694/commits/40287588c7fb79c93e45d17e6ca4dedb69119d52. We now use `labels.Labels` in both `loghttp.Entry` and `logproto.Entry`.

I wonder if we should take care of duplicated labels. I think may not (?) provided the push payload in json is an object so it should have duplicated keys (is that somehow enforced?).

Note that [proto3 supports maps][1]. In the wire, the map is the same as an array so it needs extra logic to have the map API. I ran a benchmark and looks like maps compared to arrays are ~x3 slower to marshal and ~x2 slower to unmarshal (happy to share/discuss the benchmark). Therefore I agree the way to go is using arrays in protobuf.

[1]: https://protobuf.dev/programming-guides/proto3/#maps